src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Download/PackageDownloadRunner.cs

```diff
@@ -0,0 +1,285 @@
+                            token);
+
+                    if (version == null)
+                    {
+                        // Unable to find a valid version
+                        downloadedAllSuccessfully &= false;
+                        continue;
+                    }
+
+                    bool success = await DownloadPackageAsync(
+                                        package.Id,
+                                        version,
+                                        downloadRepository,
+                                        cache,
+                                        settings,
+                                        outputDirectory,
+                                        logger,
+                                        token);
+
+                    if (success)
+                    {
+                        logger.LogMinimal(string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.PackageDownloadCommand_Succeeded,
+                            package.Id,
+                            version,
+                            outputDirectory));
+                    }
+                    else
+                    {
+                        logger.LogError(string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.PackageDownloadCommand_Failed,
+                            package.Id,
+                            version));
+
+                        downloadedAllSuccessfully &= false;
+                    }
+                }
+#pragma warning disable CA1031 // Do not catch general exception types
+                catch (Exception ex)
+                {
+                    logger.LogError(ex.ToString());
+                    downloadedAllSuccessfully &= false;
+                }
+#pragma warning restore CA1031 // Do not catch general exception types
+            }
+
+            return downloadedAllSuccessfully ? ExitCodeSuccess : ExitCodeError;
+        }
+
+        internal static async Task<(NuGetVersion, SourceRepository)> ResolvePackageDownloadVersion(
+            PackageWithNuGetVersion packageWithNuGetVersion,
+            IEnumerable<SourceRepository> sourceRepositories,
+            SourceCacheContext cache,
+            ILoggerWithColor logger,
+            bool includePrerelease,
+            CancellationToken token)
+        {
+            NuGetVersion versionToDownload = null;
+            SourceRepository downloadSourceRepository = null;
+            bool versionSpecified = packageWithNuGetVersion.NuGetVersion != null;
+
+            foreach (var repo in sourceRepositories)
+            {
+                var finder = await repo.GetResourceAsync<PackageMetadataResource>(token);
+                var packages = await finder.GetMetadataAsync(
+                    packageWithNuGetVersion.Id,
+                    includePrerelease,
+                    includeUnlisted: false,
```

Comment:
So an interesting consideration here would be, do we want to download explicitly requested packages that have been unlisted. 

I think the answer should be yes, we should download it if someone asks for it by specifying abc@version, but not allow it's dependent on latest. 

Basically disallowing unlisted versions altogether may lead to some breaks for customers because someone unlisted a package. 
If they had latest and it got unlisted, they can always fallback to specific version, but if they were depending on a specific version, then they're stuck.